### PR TITLE
Adding Battery, Battery Provider, and Battery Provider Manager

### LINKED ISSUE #42: Support org.bluez.Battery1 interface
https://github.com/bluez/bluez/blob/master/doc/battery-api.txt

### PR DESCRIPTION
**Pull request by RossComputerGuy**
Fixes #42

### CODE REVIEW

**Review by robert-ancell — CHANGES_REQUESTED** [review]
I'll re-review once the same structural changes in the BtLE PR have been made.

**Comment by RossComputerGuy** [comment]
@robert-ancell Rebased + added some changes you suggested in the BtLE PR.

File: lib/src/bluez_client.dart [hunk 1]
```diff
@@ -179,6 +180,11 @@ class BlueZAdapter {
     return _advertisingManager!;
   }
 
+  BlueZBatteryProviderManager get batteryProviderManager {
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:183]: Missing comment

File: lib/src/bluez_client.dart [hunk 2]
```diff
@@ -657,6 +663,228 @@ class BlueZAdvertisement extends DBusObject {
...
+  final BlueZClient _client;
+  final _BlueZObject _object;
+  int _nextBatteryProviderId;
+  int _nextBatteryId;
+
+  BlueZBatteryProviderManager(this._client, this._object)
+      : _nextBatteryProviderId = 0,
+        _nextBatteryId = 0;
+
+  Future<BlueZBattery> registerBattery(String? name) async {
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:680]: Missing documentation (and others below).

**Review by robert-ancell — CHANGES_REQUESTED** [review]
With the documentation fixes and some tests in `test/bluez_test.dart` this looks good, thanks!

File: lib/src/bluez_client.dart (hunk 1 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:184]: Added comment.

File: lib/src/bluez_client.dart (hunk 2 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:682]: Add documentation.

**Comment by RossComputerGuy** [comment]
I added an assert to prevent `percentage` in the battery class to be set outside the limit.

File: lib/src/bluez_client.dart [hunk 3]
```diff
@@ -657,6 +664,247 @@ class BlueZAdvertisement extends DBusObject {
...
+  final _BlueZObject _object;
+  int _nextBatteryProviderId;
+  int _nextBatteryId;
+
+  BlueZBatteryProviderManager(this._client, this._object)
+      : _nextBatteryProviderId = 0,
+        _nextBatteryId = 0;
+
+  /// Registers a new battery with an optional label.
+  Future<BlueZBattery> registerBattery(String? name) async {
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:682]: The documentation says it's a "label", the argument is "name" and the batter object has it as "source" - I would expect these to all be the same. The documentation would then use `[name]` or similar to link to that arg.

Comment (robert-ancell) [reply lib/src/bluez_client.dart:682]: "source" seems like the appropriate term after reading https://github.com/hadess/bluez/blob/master/doc/battery-api.txt

File: lib/src/bluez_client.dart [hunk 4]
```diff
@@ -657,6 +664,247 @@ class BlueZAdvertisement extends DBusObject {
...
+    return device;
+  }
+
+  /// Unregisters the battery
+  Future<void> unregisterBattery(BlueZBattery device) async {
+    await _client._bus.unregisterObject(device);
+  }
+
+  /// Registers a battery provider. A registered battery provider can
+  /// then expose objects with org.bluez.BatteryProvider(5) interface.
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:698]: The documentation talks about the DBus interfaces, which doesn't seem appropriate to this library (i.e. this is an internal detail). Reading https://github.com/hadess/bluez/blob/master/doc/battery-api.txt it's really unclear to me when you use registerBattery and when you use registerBatteryProvider - do you understand the difference? Can the documentation be updated to reflect how these functions are to be used?

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Add tests please :)

File: lib/src/bluez_client.dart (hunk 4 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:698]: `registerBattery` is used before `registerBatteryProvider` so you can provide the battery you register.

File: lib/src/bluez_client.dart (hunk 3 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Done, source is used now.

Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Added `[x]` for linking docs

File: lib/src/bluez_client.dart (hunk 4 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Clarified in doc comments.

**Comment by RossComputerGuy** [comment]
I added a test

**Comment by robert-ancell** [comment]
I had a go at testing this and improving the tests (they're not currently checking if anything is actually set when using the API). Digging into this more the current implementation is not using the BlueZ API right. I was looking at the https://github.com/bluez/bluez/blob/master/test/example-battery-provider and there should be no separate Battery object - everything is done using the org.bluez.BatteryProvider1 interface (it's in the docs but not very clear). The 'Device' property is the path to the device on the bluez server, not the object registered by the client.

The API should be something like this:
```
var client = BlueZClient();
await client.connect();
var adapter = client.adapters.first;
var device = client.devices.first;
var bpm = adapter.batteryProviderManager;
var provider = await bpm.registerBatteryProvider()
var battery = provider.addBattery(device: device, percentage: 80, source: 'Dummy Battery');
battery.percentage = 95;
provider.removeBattery(battery)
```

Note the provider also needs to set isObjectManager to true so the battery objects are announced as they are added / removed.

I had a go at updating this but I've run out of time - I'm happy to complete the test work if you are able to do the rest though.

**Comment by RossComputerGuy** [comment]
> Note the provider also needs to set isObjectManager to true so the battery objects are announced as they are added / removed.
> 
> I had a go at updating this but I've run out of time - I'm happy to complete the test work if you are able to do the rest though.

Thank you for providing this, I'll work through this when I can. Hopefully will be able to update this PR before Saturday.

**Comment by RossComputerGuy** [comment]
I had to adjust the API a bit but I believe we're closer now.

File: example/battery.dart [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  var adapter = client.adapters.first;
+  var bpm = adapter.batteryProviderManager;
+
+  var provider = await bpm.registerBatteryProvider();
+  var battery =
+      await provider.addBattery(percentage: 80, source: 'Dummy Battery');
+
+  battery.percentage = 10;
+
+  while (true) {}
```
Comment (robert-ancell) [inline example/battery.dart:16]: This line shouldn't be required - the Dart process stays open until the BlueZ client is closed, so in the example it never occurs.

**Comment by robert-ancell** [comment]
I'll add some more detail to the tests tomorrow, but I think this is pretty close to landing now.

File: example/battery.dart (hunk 5 above)
Comment (RossComputerGuy) [reply example/battery.dart:16]: The while loop line or the battery percentage line?

Comment (robert-ancell) [reply example/battery.dart:16]: The while loop.

Comment (robert-ancell) [reply example/battery.dart:16]: Having the loop like this means it sits there using a lot of CPU, without it the process is just blocking for socket input and closes when the last socket closes.

Comment (RossComputerGuy) [reply example/battery.dart]: Gotcha, GitHub didn't quite make things clear. I've removed the line.

**Comment by robert-ancell** [comment]
I think with the changes in https://github.com/MidstallSoftware/bluez.dart/pull/2 this should be good to go.

**Review by robert-ancell — APPROVED** [review]
Thanks for your patience!

**Pull request merged**